### ISSUE TITLE
I3Tree<I3Particle> cannot be deserialized (Trac #1013)

### ISSUE THREAD

**Issue by icecube-trac** [icecube-trac/tix2#1007]
Using the current icerec trunk I get the following error

"FATAL (Tree): Assertion failed: insertResult.second (I3MCTree_impl.h:625 in void TreeBase::Tree<T, Key, Hash>::insert(const T&) [with T = I3Particle, Key = I3ParticleID, Hash = __gnu_cxx::hash<I3ParticleID>])

Uncaught exception in event loop:

Assertion failed: insertResult.second (in void TreeBase::Tree<T, Key, Hash>::insert(const T&) [with T = I3Particle, Key = I3ParticleID, Hash = __gnu_cxx::hash<I3ParticleID>])"

when trying to deserialize an I3Tree<I3Particle>-object with dataio, dataio-pyshovel and steamshovel. An example file can be found here /data/user/lraedel/shared/tree_serialization.i3.bz2 It only contains a single key with the mentioned type.

Migrated from https://code.icecube.wisc.edu/ticket/1013
```json
{
    "status": "closed", 
    "changetime": "2015-08-10T20:31:17", 
    "description": "Using the current icerec trunk I get the following error\n\n\"FATAL (Tree): Assertion failed: insertResult.second (I3MCTree_impl.h:625 in void TreeBase::Tree<T, Key, Hash>::insert(const T&) [with T = I3Particle, Key = I3ParticleID, Hash = __gnu_cxx::hash<I3ParticleID>])\n\nUncaught exception in event loop:\n\nAssertion failed: insertResult.second (in void TreeBase::Tree<T, Key, Hash>::insert(const T&) [with T = I3Particle, Key = I3ParticleID, Hash = __gnu_cxx::hash<I3ParticleID>])\"\n\nwhen trying to deserialize an I3Tree<I3Particle>-object with dataio, dataio-pyshovel and steamshovel. An example file can be found here /data/user/lraedel/shared/tree_serialization.i3.bz2 It only contains a single key with the mentioned type.", 
    "reporter": "lraedel", 
    "cc": "david.schultz@icecube.wisc.edu", 
    "resolution": "fixed", 
    "_ts": "1439238677802127", 
    "component": "combo core", 
    "summary": "I3Tree<I3Particle> cannot be deserialized", 
    "priority": "critical", 
    "keywords": "", 
    "time": "2015-06-08T16:18:21", 
    "milestone": "", 
    "owner": "david.schultz", 
    "type": "defect"
}
```

**Comment by icecube-trac** [comment]
Trac update at 20150709T22:31:48: 
* david.schultz changed owner from "" to "david.schultz"
* david.schultz changed milestone from "" to "offline V15-07-00"
* david.schultz changed status from "new" to "assigned"
* david.schultz changed priority from "normal" to "critical"

**Comment by icecube-trac** [comment]
Trac update at 20150710T09:47:46: jvansanten commented:

> It looks like this happens because the particles in the tree do not have unique IDs. Here's what's tripping the assert():
> 
> ```text
> (gdb) p *head_
> $8 = (const I3ParticleID &) @0x17df318: {
>   majorID = 9681856778231206879, 
>   minorID = 71418
> }
> (gdb) p node2.GetID()
> $9 = {
>   majorID = 9681856778231206879, 
>   minorID = 71418
> }
> ```
> 
> You can end up in this situation if you initialize everything by copying a single particle. The uniqueness of keys should certainly be enforced earlier, and not just at deserialization time.

**Comment by icecube-trac** [comment]
Trac update at 20150710T10:03:48: jvansanten commented:

> This is an old-style MCTree written with dataclasses V13-12-00, which had no way of enforcing ID uniqueness, because it didn't know about IDs. What to do in this case is a fascinating question. Probably the least bad thing to do is to reassign new, unique IDs, on the theory that the tree was probably filled from the top down, and the clones of previous nodes *should* have been given unique IDs in the first place. This would orphan any I3MCHits that may have been attributed to the clones, however. It's not clear to me whether or not this is a terrible thing, since they were nonsense in the first place, but at least they pointed to somewhere.

**Comment by icecube-trac** [comment]
Trac update at 20150710T15:16:37: david.schultz commented:

> Should this be solved with an automated redoing of the particle ID when loading, like is done for dark particles?  Or maybe a standalone script that "fixes up" old data using an old version of the tree? Or just wontfix?

**Comment by icecube-trac** [comment]
Trac update at 20150710T15:26:22: olivas commented:

> I think it's safe to just treat this like Dark particles and just reissue new IDs.  As Jakob pointed out we might lose the connection to I3MCHits, but we shouldn't have many datasets around that kept I3MCHits, right?  We could also try to be smart about which ID gets reissued depending on which generates light.  If they both do, then...well really the connection was lost when the dataset was generated and there's no way to resolve that ambiguity anyway.
> 
> Whether it's a standalone script or in the serialize method and happens on-the-fly...I don't know yet.

**Comment by icecube-trac** [comment]
Trac update at 20150710T15:33:13: olivas commented:

> We definitely should not close this as "wontfix."  We've never knowingly broke backwards compatibility and have always tried to guarantee the ability to read old data with new software.  We should be able to provide some kind of solution, even if it's not perfect.

**Comment by icecube-trac** [comment]
Trac update at 20150720T16:43:33: 
* david.schultz changed cc from "" to "david.schultz@icecube.wisc.edu"
* david.schultz changed owner from "david.schultz" to "olivas"
* david.schultz commented:

> Poking olivas to make up his mind on where this should be fixed (standalone script vs inside serialization).
> 
> Feel free to reassign the ticket to me once a decision is reached.

**Comment by icecube-trac** [comment]
Trac update at 20150720T16:45:44: 
* olivas commented:

> In the serialization method on-the-fly.  I'm feeling a little more decisive this week.


* olivas changed owner from "olivas" to "david.schultz"

**Comment by icecube-trac** [comment]
Trac update at 20150720T16:52:39: olivas commented:

> Given two particles with the same PID, choose to re-issue a new ID for the particle that does *not* emit light.  This will at least preserve the connection to hits.
> 
> This needs to be advertise pretty prominently too, since any bookers that stored PIDs in ROOT files or hdf5 files will probably need to be re-generated.

**Comment by icecube-trac** [comment]
Trac update at 20150720T17:20:54: david.schultz commented:

> In this case, all particles have the same ID.  And they are all primaries. I believe this is a case of using a tree inappropriately, since the key actually says 'SplineMPEICMuEXDifferential_list'.
> 
> I'm going to go with the simple option of just giving duplicates a new id. If we can get an example with a real tree, then I'll tackle the "light" case.  It's significantly more work though, since you have to first deserialize the entire tree (giving new ids to all duplicates), then go back after the fact to look at which ones had the light and reassign the original id.

**Comment by icecube-trac** [comment]
Trac update at 20150720T17:45:07: olivas commented:

> Sounds good.

**Comment by icecube-trac** [comment]
Trac update at 20150720T18:45:31: 
* david.schultz changed resolution from "" to "fixed"
* david.schultz changed status from "assigned" to "closed"
* david.schultz commented:

> In changeset 135347:
> 
> ```text
> #!CommitTicketReference repository="IceCube" revision="135347"
> allow reading old trees with duplicate IDs. fixes #1007
> ```

**Comment by icecube-trac** [comment]
Trac update at 20150810T20:31:17: 
* olivas changed milestone from "offline V15-07-00" to ""
* olivas commented:

> Milestone offline V15-07-00 deleted
